### ISSUE TITLE
Construction of `AccessibleObject` forces creation of `Handle`

### ISSUE THREAD

**Issue by RussKie** [dotnet/winforms#3062]
* .NET Core Version: 
.NET 5 P4

**Problem description:**

Multiple code paths involving a construction or an interaction with an `AccessibleObject` on a control that hasn't had its `Handle` created cause the creation of the `Handle`.
This may lead to undesired side-effects like described in https://github.com/dotnet/winforms/issues/2797.

**Expected behavior:**

None of `AccessibleObject` events must be raised if an owner control doesn't have a valid `Handle.
https://github.com/dotnet/winforms/blob/b666dc7a94d8ac87a7d300cfb4fa86332fb79bae/src/System.Windows.Forms/src/System/Windows/Forms/AccessibleObject.cs#L2227-L2294

**Minimal repro:**

Run the following test:
https://github.com/dotnet/winforms/blob/b666dc7a94d8ac87a7d300cfb4fa86332fb79bae/src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ToolStripItemTests.cs#L14005-L14022

Line:14014 will cause `ToolStripItem.Handle` to be created.

```
 	System.Windows.Forms.dll!System.Windows.Forms.Control.CreateHandle() Line 4912	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.Handle.get() Line 2464	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.ControlAccessibleObject.ControlAccessibleObject(System.Windows.Forms.Control ownerControl = {System.Windows.Forms.ToolStrip}) Line 32	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.ToolStrip.ToolStripAccessibleObject.ToolStripAccessibleObject(System.Windows.Forms.ToolStrip owner = {System.Windows.Forms.ToolStrip}) Line 5016	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.ToolStrip.CreateAccessibilityInstance() Line 4985	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.AccessibilityObject.get() Line 498	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.ToolStripItem.ToolStripItemAccessibleObject.FragmentRoot.get() Line 364	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.AccessibleObject.Interop.UiaCore.IRawElementProviderFragment.FragmentRoot.get() Line 819	C#	Symbols loaded.
 	[Native to Managed Transition]		Annotated Frame
 	[Managed to Native Transition]		Annotated Frame
 	System.Windows.Forms.dll!System.Windows.Forms.AccessibleObject.RaiseAutomationEvent(Interop.UiaCore.UIA eventId = AutomationFocusChangedEventId) Line 2267	C#	Symbols loaded.
>	System.Windows.Forms.dll!System.Windows.Forms.ToolStripItem.ToolStripItemAccessibleObject.RaiseFocusChanged() Line 481	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.ToolStripItem.Select() Line 3094	C#	Symbols loaded.
 	System.Windows.Forms.Tests.dll!System.Windows.Forms.Tests.ToolStripItemTests.ToolStripItem_Select_InvokeWithParent_Success() Line 14017	C#	Symbols loaded.
```

**Comment by RussKie** [comment]
We could, for example, add the following guard and then run tests:
```diff
diff --git a/src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs b/src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs
index b1bb53959..fa1699363 100644
--- a/src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs
+++ b/src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs
@@ -29,6 +29,12 @@ namespace System.Windows.Forms
             public ControlAccessibleObject(Control ownerControl)
             {
                 Owner = ownerControl ?? throw new ArgumentNullException(nameof(ownerControl));
+
+                if (!ownerControl.IsHandleCreated)
+                {
+                    Debug.Fail("Accessibility object must not be created for a control without a valid handle");
+                }
+
                 IntPtr handle = ownerControl.Handle;
                 Handle = handle;
             }
```


This guard will crash the test runner, highlighting the paths where we fire off accessibility events on controls without valid handles.

Would this be a correct approach?

**Comment by hughbe** [comment]
I'm curious why I can reproduce this test failure but it doesn't repro on the CI? Any ideas

**Comment by RussKie** [comment]
No, and this has been bugging me for months. I get about a dozen of failures when run tests locally, and those are consistent (clipboard, monthcalendar dimensions, axhost...).
There must be some subtle nuances that make-or-break the tests, that we're yet to identify.

**Comment by weltkante** [comment]
If the local failures are deterministic you should be able to put additional logging into the test, dumping all relevant conditions leading to the failure and at last step fail the test intentionally. Put it in a PR to get CI execute it and look at the log to determine which of the conditions differs to local execution. Add static variables to help logging if necessary, this is just diagnostic and not supposed to be merged.

I can take a look if you don't beat me to it. Feel free to ping me on any other instances of this kind if you want me to look at, as long as they are deterministic local failures it should be possible to figure out the differences.

**Comment by hughbe** [comment]
Looking at the stack trace, it could be the implementation of `RaiseAutomationEvent`. Perhaps `UiaCore.UiaClientsAreListening` returns `true` locally but not on the CI?
```
        internal bool RaiseAutomationEvent(UiaCore.UIA eventId)
        {
            if (UiaCore.UiaClientsAreListening().IsTrue())
            {
                HRESULT result = UiaCore.UiaRaiseAutomationEvent(this, eventId);
                return result == HRESULT.S_OK;
            }

            return false;
        }
```

**Comment by weltkante** [comment]
I'm a bit confused, has the original issue already been fixed on master? `UiaClientsAreListening` is true for me but no handles are getting created in `ToolStripItem_Select_InvokeWithParent_Success`. No test failure for me locally.

**Comment by RussKie** [comment]
> I'm a bit confused, has the original issue already been fixed on master? `UiaClientsAreListening` is true for me but no handles are getting created in `ToolStripItem_Select_InvokeWithParent_Success`. No test failure for me locally.

I've created this issue after #2797 to alert @M-Lipin @vladimir-krestov @Ryuugamine (who work on accessibility implementations), to ensure our implementations don't cause a repeat, and we actively test to verify that.

**Comment by RussKie** [comment]
Copying from https://github.com/dotnet/winforms/pull/3308#issuecomment-632522821

I think these are the culprits:
https://github.com/dotnet/winforms/blob/c4e9accbc563541f17dc5ee3f9dc9ee11ce11d5f/src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs#L27-L42

I think it should be something like this:
```diff
diff --git a/src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs b/src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs
index 2f482fe43..9280a3831 100644
--- a/src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs
+++ b/src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs
@@ -4,6 +4,7 @@
 
             public ControlAccessibleObject(Control ownerControl)
             {
-                Owner = ownerControl ?? throw new ArgumentNullException(nameof(ownerControl));
-                IntPtr handle = ownerControl.Handle;
-                Handle = handle;
+                Init(ownerControl);
             }
 
             internal ControlAccessibleObject(Control ownerControl, int accObjId)
             {
-                Debug.Assert(ownerControl != null, "Cannot construct a ControlAccessibleObject with a null ownerControl");
-
                 AccessibleObjectId = accObjId; // ...must set this *before* setting the Handle property
+                Init(ownerControl);
+            }
+
+            private void Init(Control ownerControl)
+            {
                 Owner = ownerControl ?? throw new ArgumentNullException(nameof(ownerControl));
-                IntPtr handle = ownerControl.Handle;
-                Handle = handle;
+                Handle = ownerControl.IsHandleCreated ? ownerControl.Handle : IntPtr.Zero;
             }
```

BUT! Then `Handle.set` needs to be re-written, and I don't know this area sufficiently deep to suggest how. 

I tried the following very naive change, and it broke ~150 accessibility-related tests (but it didn't hang!)
```diff
             public IntPtr Handle
             {
-                get => _handle;
+                get => Owner.IsHandleCreated ? Owner.Handle : _handle;
                 set
                 {
                     if (_handle == value)
```

**Comment by vladimir-krestov** [comment]
We don't need to change `Handle.get` because `_handle` is set when creating control `Handle` (`OnHandleCreated` method).
And if `Handle` is created then (`Owner.Handle  ==  _handle`).
